### DISCUSSION TITLE
How to estimate the average branching factor for code generation with grammar guided generation?

### DISCUSSION THREAD

**Discussion by arunpatro** [outlines-dev/outlines#554] [General]
Naive code generation has a branching factor of about 50K. Since, outlines uses guided generation and generates a NFA, is there a way to observe what and how many logits are valid? I want to be able to calculate the average branching factor for next token generation. 

Can somehow show me how to calculate the avg. no. of valid next tokens per token for a sequence when provided a grammar?

**Comment by rlouf** [comment]
A good start would be to look at [CFGFSM](https://github.com/outlines-dev/outlines/blob/65ed0f7e9f900701ba62479982ee29bb47cc2738/outlines/fsm/fsm.py#L214). You can run it manually on a given piece of code.
